parquet/src/main/java/org/apache/iceberg/parquet/ParquetReader.java

```diff
@@ -117,7 +122,12 @@ private static class FileIterator<T> implements CloseableIterator<T> {
 
     @Override
     public boolean hasNext() {
-      return valuesRead < totalValues;
+      long numToRead = totalValues;
+      if (pushedLimit > 0 && pushedLimit < numToRead) {
```

Comment:
The special casing of -1/0 makes our logic more complicated here, maybe we should reconsider?